### ISSUE TITLE
v5.21.6-336-gfedf30e introduced a memory leak

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14541]
Migrated from [rt.perl.org#123922](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123922) (status was 'resolved')

Searchable as RT123922$

**Comment by p5pRT** [comment]
### From @tonycoz

#### Created by @tonycoz



fedf30e1c349130b23648c022f5f3cb4ad7928f3 \(v5\.21\.6\-336\-gfedf30e\)
introduced a memory leak when running the following code&#8203;:

PERL\_DESTRUCT\_LEVEL=2 valgrind \-q \-\-leak\-check=full \./perl \-Ilib \-Mwarnings \-e0

In blead the reported leaks are&#8203;:

==23030== 3\,584 \(512 direct\, 3\,072 indirect\) bytes in 1 blocks are definitely lost in loss record 12 of 26
==23030==    at 0x4C267BB&#8203;: calloc \(vg\_replace\_malloc\.c&#8203;:593\)
==23030==    by 0x4203A6&#8203;: Perl\_Slab\_Alloc \(op\.c&#8203;:181\)
==23030==    by 0x428C07&#8203;: Perl\_newSVOP \(op\.c&#8203;:5807\)
==23030==    by 0x46F985&#8203;: S\_force\_word \(toke\.c&#8203;:2034\)
==23030==    by 0x485493&#8203;: Perl\_yylex \(toke\.c&#8203;:7739\)
==23030==    by 0x48D5DD&#8203;: Perl\_yyparse \(perly\.c&#8203;:322\)
==23030==    by 0x58B9A7&#8203;: S\_doeval \(pp\_ctl\.c&#8203;:3484\)
==23030==    by 0x5929B9&#8203;: Perl\_pp\_require \(pp\_ctl\.c&#8203;:4169\)
==23030==    by 0x4D630A&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2231\)
==23030==    by 0x44788E&#8203;: Perl\_call\_sv \(perl\.c&#8203;:2736\)
==23030==    by 0x4482A1&#8203;: Perl\_call\_list \(perl\.c&#8203;:4842\)
==23030==    by 0x42228A&#8203;: S\_process\_special\_blocks \(op\.c&#8203;:8899\)
==23030== 
==23030== 23\,552 \(1\,024 direct\, 22\,528 indirect\) bytes in 2 blocks are definitely lost in loss record 24 of 26
==23030==    at 0x4C267BB&#8203;: calloc \(vg\_replace\_malloc\.c&#8203;:593\)
==23030==    by 0x4203A6&#8203;: Perl\_Slab\_Alloc \(op\.c&#8203;:181\)
==23030==    by 0x42A193&#8203;: Perl\_newOP \(op\.c&#8203;:4748\)
==23030==    by 0x48E642&#8203;: Perl\_yyparse \(perly\.y&#8203;:996\)
==23030==    by 0x58B9A7&#8203;: S\_doeval \(pp\_ctl\.c&#8203;:3484\)
==23030==    by 0x5929B9&#8203;: Perl\_pp\_require \(pp\_ctl\.c&#8203;:4169\)
==23030==    by 0x4D630A&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2231\)
==23030==    by 0x44788E&#8203;: Perl\_call\_sv \(perl\.c&#8203;:2736\)
==23030==    by 0x4482A1&#8203;: Perl\_call\_list \(perl\.c&#8203;:4842\)
==23030==    by 0x42228A&#8203;: S\_process\_special\_blocks \(op\.c&#8203;:8899\)
==23030==    by 0x43BE95&#8203;: Perl\_newATTRSUB\_x \(op\.c&#8203;:8829\)
==23030==    by 0x43EA7E&#8203;: Perl\_utilize \(op\.c&#8203;:6052\)
==23030== 
==23030== 48\,640 \(512 direct\, 48\,128 indirect\) bytes in 1 blocks are definitely lost in loss record 25 of 26
==23030==    at 0x4C267BB&#8203;: calloc \(vg\_replace\_malloc\.c&#8203;:593\)
==23030==    by 0x4203A6&#8203;: Perl\_Slab\_Alloc \(op\.c&#8203;:181\)
==23030==    by 0x428C07&#8203;: Perl\_newSVOP \(op\.c&#8203;:5807\)
==23030==    by 0x46F985&#8203;: S\_force\_word \(toke\.c&#8203;:2034\)
==23030==    by 0x485EFD&#8203;: Perl\_yylex \(toke\.c&#8203;:7638\)
==23030==    by 0x48D5DD&#8203;: Perl\_yyparse \(perly\.c&#8203;:322\)
==23030==    by 0x58B9A7&#8203;: S\_doeval \(pp\_ctl\.c&#8203;:3484\)
==23030==    by 0x5929B9&#8203;: Perl\_pp\_require \(pp\_ctl\.c&#8203;:4169\)
==23030==    by 0x4D630A&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2231\)
==23030==    by 0x44788E&#8203;: Perl\_call\_sv \(perl\.c&#8203;:2736\)
==23030==    by 0x4482A1&#8203;: Perl\_call\_list \(perl\.c&#8203;:4842\)
==23030==    by 0x42228A&#8203;: S\_process\_special\_blocks \(op\.c&#8203;:8899\)
==23030== 
==23030== 64\,000 \(1\,536 direct\, 62\,464 indirect\) bytes in 3 blocks are definitely lost in loss record 26 of 26
==23030==    at 0x4C267BB&#8203;: calloc \(vg\_replace\_malloc\.c&#8203;:593\)
==23030==    by 0x4203A6&#8203;: Perl\_Slab\_Alloc \(op\.c&#8203;:181\)
==23030==    by 0x42A193&#8203;: Perl\_newOP \(op\.c&#8203;:4748\)
==23030==    by 0x4722F9&#8203;: Perl\_yylex \(toke\.c&#8203;:8243\)
==23030==    by 0x48D5DD&#8203;: Perl\_yyparse \(perly\.c&#8203;:322\)
==23030==    by 0x58B9A7&#8203;: S\_doeval \(pp\_ctl\.c&#8203;:3484\)
==23030==    by 0x5929B9&#8203;: Perl\_pp\_require \(pp\_ctl\.c&#8203;:4169\)
==23030==    by 0x4D630A&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2231\)
==23030==    by 0x44788E&#8203;: Perl\_call\_sv \(perl\.c&#8203;:2736\)
==23030==    by 0x4482A1&#8203;: Perl\_call\_list \(perl\.c&#8203;:4842\)
==23030==    by 0x42228A&#8203;: S\_process\_special\_blocks \(op\.c&#8203;:8899\)
==23030==    by 0x43BE95&#8203;: Perl\_newATTRSUB\_x \(op\.c&#8203;:8829\)
==23030== 

Simple code such as&#8203;:

$ valgrind \-q \-\-leak\-check=full \./perl \-e '$x\->\[1\]\[2\]' 
==23620== 1\,536 \(512 direct\, 1\,024 indirect\) bytes in 1 blocks are definitely lost in loss record 3 of 3
==23620==    at 0x4C267BB&#8203;: calloc \(vg\_replace\_malloc\.c&#8203;:593\)
==23620==    by 0x4203A6&#8203;: Perl\_Slab\_Alloc \(op\.c&#8203;:181\)
==23620==    by 0x428C07&#8203;: Perl\_newSVOP \(op\.c&#8203;:5807\)
==23620==    by 0x473AB6&#8203;: Perl\_yylex \(toke\.c&#8203;:8308\)
==23620==    by 0x48D5DD&#8203;: Perl\_yyparse \(perly\.c&#8203;:322\)
==23620==    by 0x44DEF5&#8203;: S\_parse\_body \(perl\.c&#8203;:2277\)
==23620==    by 0x44F3AB&#8203;: perl\_parse \(perl\.c&#8203;:1611\)
==23620==    by 0x41B971&#8203;: main \(perlmain\.c&#8203;:114\)
==23620== 

demonstrates one point of leakage\, I haven't tried to duplicate the
others\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.21.10:

Configured by tony at Wed Feb 25 15:00:37 AEDT 2015.

Summary of my perl5 (revision 5 version 21 subversion 10) configuration:
  Commit id: 33593911f214382b592d05aa902655301915e666
  Platform:
    osname=linux, osvers=2.6.32-358.el6.x86_64, archname=x86_64-linux-thread-multi
    uname='linux dromedary-001.ams6.corp.booking.com 2.6.32-358.el6.x86_64 #1 smp fri feb 22 00:31:26 utc 2013 x86_64 x86_64 x86_64 gnulinux '
    config_args='-Dprefix= -Uversiononly -Dusedevel -des -Dusethreads -DDEBUGGING=both'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.7 20120313 (Red Hat 4.4.7-11)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /lib/../lib64 /usr/lib/../lib64 /lib /lib64 /usr/lib64 /usr/local/lib64
    libs=-lpthread -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.12.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.12'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib -fstack-protector'


Characteristics of this binary (from libperl): 
  Compile-time options: DEBUGGING HAS_TIMES MULTIPLICITY PERLIO_LAYERS
                        PERL_DONT_CREATE_GVSV
                        PERL_HASH_FUNC_ONE_AT_A_TIME_HARD
                        PERL_IMPLICIT_CONTEXT PERL_MALLOC_WRAP
                        PERL_NEW_COPY_ON_WRITE PERL_PRESERVE_IVUV
                        PERL_TRACK_MEMPOOL PERL_USE_DEVEL USE_64_BIT_ALL
                        USE_64_BIT_INT USE_ITHREADS USE_LARGE_FILES
                        USE_LOCALE USE_LOCALE_COLLATE USE_LOCALE_CTYPE
                        USE_LOCALE_NUMERIC USE_LOCALE_TIME USE_PERLIO
                        USE_PERL_ATOF USE_REENTRANT_API
  Built under linux
  Compiled at Feb 25 2015 05:29:40
  @INC:
    lib
    /usr/local/lib/perl5/site_perl/5.21.10/x86_64-linux-thread-multi
    /usr/local/lib/perl5/site_perl/5.21.10
    /usr/local/lib/perl5/5.21.10/x86_64-linux-thread-multi
    /usr/local/lib/perl5/5.21.10
    /usr/local/lib/perl5/site_perl/5.12.3
    /usr/local/lib/perl5/site_perl/5.8.8
    /usr/local/lib/perl5/site_perl
    .

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Feb 24\, 2015 at 08&#8203;:40&#8203;:33PM \-0800\, Tony Cook wrote&#8203;:
> fedf30e1c349130b23648c022f5f3cb4ad7928f3 \(v5\.21\.6\-336\-gfedf30e\)
> introduced a memory leak when running the following code&#8203;:
> 
> PERL\_DESTRUCT\_LEVEL=2 valgrind \-q \-\-leak\-check=full \./perl \-Ilib \-Mwarnings \-e0

Thanks for spotting\.
Fixed with b7613b8a45c70113cb7f32d0bcfc9a03889ae272\.

\-\- 
The warp engines start playing up a bit\, but seem to sort themselves out
after a while without any intervention from boy genius Wesley Crusher\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#17

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thank you for submitting this ticket\.

The issue should now be resolved with the release today of Perl v5\.22\, which is available at http&#8203;://www\.perl\.org/get\.html
\-\- 
Karl Williamson for the Perl 5 team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
